File: cuhara.qua.back/src/core/cqrs/query.ts
export interface Query<TResponse = void> {
  readonly type: string;
}

export interface QueryHandler<
  TQuery extends Query<TResponse>,
  TResponse = void
> {
  handle(query: TQuery): Promise<TResponse>;
}

export abstract class BaseQuery<TResponse = void> implements Query<TResponse> {
  abstract readonly type: string;
}

File: cuhara.qua.go/internal/common/cqrs/query.go
package cqrs

import "context"

type Query[T any] interface {
	Type() string
}

type QueryHandler[T any] interface {
	Handle(ctx context.Context, query Query[T]) (T, error)
}

type BaseQuery[T any] struct {
	queryType string
}

func NewBaseQuery[T any](queryType string) BaseQuery[T] {
	return BaseQuery[T]{
		queryType: queryType,
	}
}

func (c BaseQuery[T]) Type() string {
	return c.queryType
}
